### DISCUSSION TITLE
`tar_format` write & read multiple files

### DISCUSSION THREAD

**Discussion by dipterix** [ropensci/targets#985] [Ideas]
Some data are stored in multiple files. For example, bigmemory, filematrix, filearray, disk.frame... The pattern is to store large data separated from "header"s, which are used to generate small in-memroy stubs.

I was wondering how hard would it be to support `tar_format` to write the header information into `path`, in the meanwhile allowing to store massive data into a new parameter `path_data` (which is a folder)? Since the header can contain all the information (including the signature of the data), you only need to hash the headers.

Another use case would be if my target is an environment that is generated from parameters, I can save the environment into `path_data`, and the generating parameters into `path`. Such way provides better cache (as saving the environment will likely to trigger an update, but saving the parameters are less likely..., even under `cue="always"`)

As for `path_data`, you can let it be `${path}+data`. for example, if `path` is `xxx/obj_name0afb23`, then `path_data` can be `xxx/obj_name0afb23+data` as a folder. If the folder exists, you can just simply copy it over to the `shared` folder (and rename to `obj_name+data`

**Comment by wlandau** [comment]
`tar_format()` is not really in a position to get that complicated. For the cases you describe, you could use a `format = "file"` target and only supply the path to the header file. Then you could write a simple function to quickly reconstruct the object from that header file and call it from the targets that depend on it.

**Reply by dipterix** [reply]
Gotcha, so is it safe that I save the data to `shared/objects/` by myself manually? Right now I dynamically create format function to be able to pass in the target name, using `targets::tar_path`, I can get the shared object path, and by manually creating `+data` folder, I was able to write my data. 

If `targets` is not gonna support the head-data format, I would like to make sure that what I'm doing is legit to you so the future updates won't (or are unlikely to) break this.

My data repository is always local (no remotes like aws)

**Reply by wlandau** [reply]
Seems like your approach should work. However, I do suggest you write the small tracked file to _targets/objects/ and the large untracked data files to _targets/user/. If you write a special +data folder to _targets/objects, there may be unforeseen consequences depending on future development in targets.  

Either way, tar_delete() will not remove the large data files, and there is an implicit assumption that the large data files are always synchronized with the header file. (If a large data file gets corrupted or deleted but the header file stays the same, then tar_make() is oblivious.)

**Reply by dipterix** [reply]
Thanks. This is great to know I should store data in _targets/user/
